### ISSUE TITLE
Improve the process of fine-tuning buildSrc

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#6045]
When a change is made to anything in `buildSrc` every task (except ones shipped with Gradle) lose their caching. This is to be expected, as the plugin classpath changes, but it has some problematic consequences:

**For CI and developers**

When a change to `buildSrc` is pushed, the first CI build to follow will likely take a lot longer than usual. Later CI builds should be able to rely on the remote cache, so they are not affected. Developers though will lose most of the local incremental state, and most of their tasks that would otherwise be `UP-TO-DATE` need to pull changes from the remote cache instead, or be re-executed if they are not cacheable.

**For the build author**

While a build author is working on making a change to `buildSrc`, it is typical that they want to repeatedly try out iterative changes. Without an already populated remote cache to rely upon, they must wait for all third-party and custom tasks to execute every time they make a change every time they want to test a new version. This can slow down `buildSrc` development significantly, especially when it comes to changes that affect behavior towards the end of a long build.

While the CI/developer case is acceptable in most cases, it makes sense to offer ways to improve the second.

Build authors can already employ tricks with selectively excluding some tasks (`-exclude-task`), and by not rebuilding dependencies (`--no-rebuild`). However, the latter option is already deprecated, and the first one is not very handy in a large project.

Another option is to manually remove dependencies in the build to separate the task(s) that are to be tested. This is of course not easy to do in some cases, and is pretty error-prone, too.

The ultimate way to fix all this would be to precisely detect the code paths accessible by each task, and only invalidate the ones that actually refer to the changed code. This would also fix use cases where upgrading one plugin leads to the invalidation of tasks created by completely unrelated plugins. This solution also sounds pretty sci-fi, and has the potential of leading to OSGi.

Anyway, until we have something super-duper like selective access detection, we can already take the existing tools and make them into something that is a) not deprecated, b) support testing changes to `buildSrc` by ignoring the classpath changes where it can be (relatively) safely ignored based on the user's input.

**Comment by wolfs** [comment]
Let's not forget about an already existing way to quickly try out changes in buildSrc: add some functional tests for plugins/tasks defined in `buildSrc`.

**Comment by adammurdoch** [comment]
A better (cheaper) increment would be to better separate the classpaths/loaders of the various buckets of build logic. There's no good reason, for example, for an external plugin to see any of the classes defined in `buildSrc`, and only downsides. So, we might split these into two separate classpaths. A similar approach would be to help people move their local build logic from `buildSrc` to child included builds so that a given implementation class is visible only in those projects that use the local plugin and are not visible to external plugins. This isn't the same as fine grained per-plugin isolation (which would be the goal). Instead of one giant bucket, there'd be 2 buckets, one for external stuff and one for local stuff. Then I can change the local stuff without invalidating any of the external stuff.

We might do a similar thing for external plugins that are included by specific build scripts, and then increment to finer grained isolation.

**Comment by ldaley** [comment]
> There's no good reason, for example, for an external plugin to see any of the classes defined in buildSrc, and only downsides. 

I have abused this before to patch a plugin. Not saying it's a _good_ reason.

**Comment by oehme** [comment]
I think we should do something about this for 5.0, because

- We're making the Kotlin DSL 1.0 and its precompiled scripts utilize buildSrc
- We're deprecating `--no-rebuild`, which is the only current workaround

Something simple like Adam proposed above would fix the situation for most users.

**Comment by wolfs** [comment]
@oehme Is this still scheduled for Gradle 5.0?

**Comment by oehme** [comment]
Yes, as mentioned above I believe we should do something for 5.0.

**Comment by marcphilipp** [comment]
@gradle/build-cache Do you have any plans to work on this issue in the 5.0 timeframe?

**Comment by marcphilipp** [comment]
We'll undeprecated `--no-rebuild` for now and document its usefulness when working on `buildSrc` in the user guide in #6968.

**Comment by jbytecoder** [comment]
I would like to note that if the buildLogic classlaoder could be split,  the buildScript should receive it's own classloader independent of buildSrc. This way if I have a task with ```doLast { project.delete("stuff") }```
it could not be invalidated by changes to buildSrc

**Comment by oehme** [comment]
> the buildScript should receive it's own classloader independent of buildSrc

The whole purpose of `buildSrc` is to provide logic that build scripts can access. Therefor it must be a parent for them.

**Comment by jbytecoder** [comment]
@oehme  sorry I did not express my self precisely

What I meant with independent is that there would be 2 classLoader one for buildSrc and one for compiled build.gradle ( obviously the first would be the parent of the second ) with
diffrent, non overlapping paths - independent paths 

All ```buildscript {}``` dependencies could be relay attached to buildSrc classloader, while the only thing on the second classloader would be classes that resulted from compilation of build.gradle. 

This way it would be clear that task actions that do not reference classes from buildSrc change only when build.gradle changes.  It would also mean that when you update a ```buildscript {}``` dependency it would not invalidate actions that only use the gradle api.  

In any case this idea makes sens only if it's relatively easy to determine whether a Closure references classes from one or multiple classloaders - I thought it would be easier than performing full analysis of depended classes and whether they where changed

**Comment by oehme** [comment]
> that task actions that do not reference classes from buildSrc change only when build.gradle changes

Knowing that would require doing bytecode analysis and even then it could break if dynamic Groovy or reflection is used.

**Comment by realdadfish** [comment]
Maybe I'm totally ignorant, but wouldn't it be possible somehow to "monitor" the API of what's in `buildSrc` (by using metalava or something similar) and only then trigger a recompilation / reconfiguration if the API changed? Then all non-API-related changes, like changes in how custom tasks are working internally, could be super quick, no?

**Comment by isker** [comment]
@realdadfish this isn’t about APIs but about underlying implementations.  API change analysis applies only for eliding unnecessary compilation of downstream code.  But code downstream of buildSrc (i.e. the rest of the build) is _running_ buildSrc at configuration time in addition to compiling against it, so API analysis would not meaningfully improve anything here.

**Comment by isker** [comment]
@realdadfish this isn’t about APIs but about underlying implementations.  API change analysis applies only for eliding unnecessary compilation of downstream code.  But code downstream of buildSrc (i.e. the rest of the build) is _running_ buildSrc at configuration time in addition to compiling against it, so API analysis would not meaningfully improve anything here.

**Comment by realdadfish** [comment]
But could at least the recompilation of build.gradle.kts files be spared then, do existing compiled files are ran against an (otherwise unchanged) buildSrc? 

Am 21. April 2020 18:27:55 MESZ schrieb Ian Kerins <notifications@github.com>:
>@realdadfish this isn’t about APIs but about underlying
>implementations.  API change analysis applies only for eliding
>unnecessary compilation of downstream code.  But code downstream of
>buildSrc (i.e. the rest of the build) is _running_ buildSrc at
>configuration time in addition to compiling against it, so API analysis
>would not meaningfully improve anything here.>
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/gradle/gradle/issues/6045#issuecomment-617265687

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by cvmocanu** [comment]
Are there any disadvantages replacing ``buildSrc`` with a composite build that includes a subproject (``includeBuild`` in ``settings.gradle.kts``) -- see https://proandroiddev.com/stop-using-gradle-buildsrc-use-composite-builds-instead-3c38ac7a2ab3?

Composite builds don't invalidate the entire build when the code in the included build changes.

If there are no disadvantages to this approach, ``buildSrc`` could be deprecated and removed in a future version.

**Comment by jjohannes** [comment]
> If there are no disadvantages to this approach, buildSrc could be deprecated and removed in a future version.

@cvmocanu that's the plan. We don't know which version that will be but we are actively working on the topic now. There are some "features" in buildSrc (see below) that won't work with included builds. You shouldn't use these features, but there are potentially many builds out there that do. We need to provide a feasible migration path before we can deprecate buildSrc (or turn it into an included build). That's the main reason why it is not deprecated yet.

Right now we are working on some composite build improvements, which will make them superior to buildSrc. E.g. you'll be able to call a task from an included build directly, for example to do `included-build:check` explicitly.

> Are there any disadvantages replacing buildSrc with a composite build 

There are some buildSrc features that do not work with composites. Like using a class directly without applying a plugin, or the fact that buildSrc reuses repository declarations from the main build. But these are all things that work historically, but shouldn't (if we would invent buildSrc today, we won't add these "features").

So if you don't need these features in your build for build historical reasons, I would recommend using an included build instead of buildSrc. Especially if you break up your build logic into multiple subprojects (or multiple included builds) you can have this effect, which you correctly stated:

> Composite builds don't invalidate the entire build when the code in the included build changes.

And you will profit from included build features in the next Gradle versions which we most likely won't "backport" to buildSrc anymore.

**Comment by daniel-jasinski** [comment]
This issue is a big pain point for us since we have a non-trivial amount of native code to build.

In my opinion, the problem arises because Gradle here tries to infer the **semantics** from the **implementation**, which is exactly the wrong thing to do.

For example, if I have my custom MyCopy task if I change the implementation from using standard Java IO to using Guava or Apache library, users of my task should not care providing I did not introduce any bugs.

Most tasks (all built-in?) have already fixed behavior (semantics) that we rely on in our builds.
Although this is the most common case, the build system cannot rely on this fact, nor can it infer it.

The semantics is only known to the task author, therefore there should be a way to forward this information to Gradle.
An example solution is to introduce an optional annotation on the task classes, like `@TaskSemantics(version = 0)`.
Then version declared in such a way can become part of the cache entry metadata. If the cache entry did not include such a version or the current version is different than the previous, then the cache is invalidated otherwise it remains valid.
This solution has a neat feature, that it is opt-in therefore allowing for gradual migration.

**Comment by aspic-fish** [comment]
@jjohannes could you clarify whether buildSrc is still going to be deprecated in the future? I've failed to find any relevant info on the internet. I suspect plans might have changed.

**Comment by jjohannes** [comment]
@aspic-fish I don't know the exact plans right now as I am no longer part of the Gradle team. But my last information is that eventually buildSrc should become an _included build_. What I explained in https://github.com/gradle/gradle/issues/6045#issuecomment-717752760 is all possible with Gradle 7.x through `pluginManagement { includeBuild("...") }` and I would always us that today and then you won't have to care about what happens to buildSrc.

There are also several related issues on this topic it seems. I think this one is most likely to get updated if there are any news: #2531.

**Comment by jbartok** [comment]
The issue is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before a fix is made.
